rasterio/_warp.pyx

```diff
@@ -487,3 +491,60 @@ def _reproject(
         if dtypes.is_ndarray(source):
             if hdsin != NULL:
                 _gdal.GDALClose(hdsin)
+
+
+def _calculate_default_transform(
+        src_crs, dst_crs, width, height, left, bottom, right, top,
+        resolution=None, densify_pts=21):
+    """Wraps GDAL's algorithm."""
+
+    cdef void *hTransformArg = NULL
+    cdef int npixels = 0
+    cdef int nlines = 0
+    cdef double extent[4]
+    cdef double geotransform[6]
+    cdef void *osr = NULL
+    cdef char *wkt = NULL
+    cdef InMemoryRaster temp = None
+
+    extent[:] = [0.0, 0.0, 0.0, 0.0]
+    geotransform[:] = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+
+    # Make an in-memory raster dataset we can pass to 
+    # GDALCreateGenImgProjTransformer().
+    transform = from_bounds(left, bottom, right, top, width, height)
+    img = np.ones((height, width))
+
+    osr = _base._osr_from_crs(dst_crs)
+    _gdal.OSRExportToWkt(osr, &wkt)
+    _gdal.OSRDestroySpatialReference(osr)
+
+    with InMemoryRaster(
+            img, transform=transform.to_gdal(), crs=src_crs) as temp:
+        hTransformArg = _gdal.GDALCreateGenImgProjTransformer(
```

Comment:
@sgillies it looks like [`gdalwarp.cpp`](https://trac.osgeo.org/gdal/browser/branches/1.11/gdal/apps/gdalwarp.cpp#L1684) is using `GDALCreateGenImgProjTransformer2` instead of `GDALCreateGenImgProjTransformer`.  Possibly simply due to fewer required arguments?  I haven't followed the code through enough to see if it depends on something you don't have at hand here, although it would mean wrapping  `GDALCreateGenImgProjTransformer2` in order to use.  If you leave as is, a comment explaining how you got the last 3 parameters may be helpful.